contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -2716,54 +2716,49 @@ exec_stmt_insert_execute_select(PLtsql_execstate *estate, PLtsql_expr *query)
 
 int exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
 {
-	char *bulk_load_schema_name = NULL;
 	MemoryContext	oldContext;
 	Oid schema_oid = InvalidOid;
 
+	oldContext = MemoryContextSwitchTo(TopMemoryContext);
+
+	/*
+	 * We use a global variable so that we do not need to call BeginBulkCopy
+	 * in case of implicit batching, which saves time.
+	 */
+	cstmt = (BulkCopyStmt *) palloc0(sizeof(BulkCopyStmt));
```

Comment:
You should allocate memory in TopMemoryContext only if you want to keep the allocations throughout the backend lifetime. The memory doesn't get reclaimed during an error.
IIUC, you want to palloc this memory for each insert bulk statement you're executing. In this case, it's better to allocate in a separate memory context that you can free after each execution (even in case of error).

Reply:
We have a copycontext which is part of BulkCopyStateData which is responsible for holding data during the actually copy/insertion.

But for your comment, we are creating cstmt (BulkCopyStmt) which communicates information from insert bulk query (sql batch req1) to the bulk load request (req2).That will require TOP Memory context to persist information outside per-query context. Even if we create a new context here, it will not give us additional benefit.

Regarding your concern on the cleanup of these allocations, please check these links to the code where we have called cleanup from TDS for error and non error scenarios:
1. Cleanup from TDS CATCH block:
[https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/651/files#diff-f01df05dcab287463fcee6b45e4[…]ba61e5c0f4ebbdb848343bdd32832R906](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/651/files#diff-f01df05dcab287463fcee6b45e49fa8f46bba61e5c0f4ebbdb848343bdd32832R906)
2. For non-error scenario we cleanup here:
[https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/651/files#diff-f01df05dcab287463fcee6b45e4[…]5c0f4ebbdb848343bdd32832R787-R789](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/651/files#diff-f01df05dcab287463fcee6b45e49fa8f46bba61e5c0f4ebbdb848343bdd32832R787-R789)

Actual cleanup:
Cleanup inside callback function:
[https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/651/files#diff-3ca2886287e484395b98c3d6f90[…]ef1e975c1bb1c79b93e7abR2789-R2792](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/651/files#diff-3ca2886287e484395b98c3d6f90cb766e4f75c5cd9ef1e975c1bb1c79b93e7abR2789-R2792)

Reply:
there are still transaction context and message context which goes beyond query. Top memory context should be last one to use